File: src/lib/job-safety.ts
import { createHash } from "node:crypto";
import safetyConfig from "@/config/job-safety.json";

export type JobSafetyDisposition = "ACCEPT" | "REVIEW" | "REJECT";
export type JobSafetyProfile = keyof typeof safetyConfig.profiles;

export interface JobSafetyResult {
  disposition: JobSafetyDisposition;
  reason: string;
  publicTitle?: string;
  profile?: JobSafetyProfile;
}

export interface RawJobIdentity {
  trade: unknown;
  id: unknown;
  jobUrl: unknown;
  source?: unknown;
}

const idPattern = new RegExp(safetyConfig.idPattern);
const standaloneEfzPattern = /(?:^|[^\p{L}\p{N}_])efz(?:$|[^\p{L}\p{N}_])/iu;

function normalize(value: unknown): string {
  return typeof value === "string"
    ? value.normalize("NFKC").toLocaleLowerCase("de-CH").replace(/\s+/g, " ").trim()
    : "";
}

function includesAny(value: string, patterns: readonly string[]): boolean {
  return patterns.some((pattern) => value.includes(pattern));
}

export function classifyZimmermannTitle(titleValue: unknown): JobSafetyResult {
  const title = normalize(titleValue);
  if (!title) return { disposition: "REJECT", reason: "missing-title" };

  if (includesAny(title, safetyConfig.hardNegativeTitlePatterns)) {
    return { disposition: "REJECT", reason: "hard-negative-title" };
  }

  const specific = safetyConfig.highSpecificityTitles.find(({ pattern }) =>
    new RegExp(pattern, "iu").test(title),
  );
  const hasTradeContext = includesAny(title, safetyConfig.tradeContextPatterns);
  const broad = hasTradeContext
    ? safetyConfig.broadTitlesRequiringTradeContext.find(({ pattern }) =>
        new RegExp(pattern, "iu").test(title),
      )
    : undefined;
  const match = specific ?? broad;

  // Descriptions are intentionally absent from this classifier. Body text can
  // never rescue a generic or unrelated title.
  if (!match) return { disposition: "REJECT", reason: "no-zimmermann-title-signal" };

  if (includesAny(title, safetyConfig.otherTradeTitlePatterns)) {
    return { disposition: "REVIEW", reason: "mixed-trade-title" };
  }

  const hasStandaloneEfz = standaloneEfzPattern.test(title);
  return {
    disposition: "ACCEPT",
    reason: specific ? "specific-zimmermann-title" : "broad-role-with-zimmermann-context",
    publicTitle: `${match.publicTitle}${hasStandaloneEfz ? " EFZ" : ""}`,
    profile: match.profile as JobSafetyProfile,
  };
}

export function validateRawJobIdentity(job: RawJobIdentity): string | null {
  if (job.trade !== safetyConfig.trade) return "wrong-trade";
  if (typeof job.id !== "string" || !idPattern.test(job.id)) return "invalid-id";
  if (typeof job.jobUrl !== "string" || !/^https?:\/\//i.test(job.jobUrl)) return "invalid-job-url";
  const expectedId = `scraped-${safetyConfig.trade}-${createHash("md5").update(job.jobUrl).digest("hex").slice(0, 12)}`;
  if (job.id !== expectedId) return "id-url-mismatch";
  if (
    job.source !== undefined &&
    (typeof job.source !== "string" || !safetyConfig.allowedSources.includes(job.source.toLowerCase()))
  ) {
    return "invalid-source";
  }
  return null;
}

export function isValidZimmermannJobId(value: unknown): value is string {
  return typeof value === "string" && idPattern.test(value);
}

export function getZimmermannProfile(profile: JobSafetyProfile) {
  return safetyConfig.profiles[profile];
}

export { safetyConfig as JOB_SAFETY_CONFIG };

File: scripts/job_safety.py
"""Shared deny-first Zimmermann classifier backed by src/config/job-safety.json."""

from __future__ import annotations

import hashlib
import json
import re
import unicodedata
from pathlib import Path
from typing import Any

CONFIG_PATH = Path(__file__).resolve().parents[1] / "src" / "config" / "job-safety.json"
CONFIG = json.loads(CONFIG_PATH.read_text(encoding="utf-8"))
ID_PATTERN = re.compile(CONFIG["idPattern"])
STANDALONE_EFZ = re.compile(r"(?:^|[^\w])efz(?:$|[^\w])", re.IGNORECASE)


def normalize(value: Any) -> str:
    if not isinstance(value, str):
        return ""
    return re.sub(r"\s+", " ", unicodedata.normalize("NFKC", value).casefold()).strip()


def classify_zimmermann_title(title_value: Any) -> dict[str, Any]:
    title = normalize(title_value)
    if not title:
        return {"disposition": "REJECT", "reason": "missing-title"}

    if any(pattern in title for pattern in CONFIG["hardNegativeTitlePatterns"]):
        return {"disposition": "REJECT", "reason": "hard-negative-title"}

    specific = next(
        (rule for rule in CONFIG["highSpecificityTitles"] if re.search(rule["pattern"], title, re.IGNORECASE)),
        None,
    )
    has_trade_context = any(pattern in title for pattern in CONFIG["tradeContextPatterns"])
    broad = None
    if has_trade_context:
        broad = next(
            (
                rule
                for rule in CONFIG["broadTitlesRequiringTradeContext"]
                if re.search(rule["pattern"], title, re.IGNORECASE)
            ),
            None,
        )
    match = specific or broad

    # Description/body text is never inspected here by design.
    if not match:
        return {"disposition": "REJECT", "reason": "no-zimmermann-title-signal"}

    if any(pattern in title for pattern in CONFIG["otherTradeTitlePatterns"]):
        return {"disposition": "REVIEW", "reason": "mixed-trade-title"}

    efz_suffix = " EFZ" if STANDALONE_EFZ.search(title) else ""
    return {
        "disposition": "ACCEPT",
        "reason": "specific-zimmermann-title" if specific else "broad-role-with-zimmermann-context",
        "publicTitle": f"{match['publicTitle']}{efz_suffix}",
        "profile": match["profile"],
    }


def validate_raw_job_identity(job: Any) -> str | None:
    if not isinstance(job, dict):
        return "invalid-job"
    if job.get("trade") != CONFIG["trade"]:
        return "wrong-trade"
    job_id = job.get("id")
    job_url = job.get("jobUrl")
    if not isinstance(job_id, str) or not ID_PATTERN.fullmatch(job_id):
        return "invalid-id"
    if not isinstance(job_url, str) or not re.match(r"^https?://", job_url, re.IGNORECASE):
        return "invalid-job-url"
    digest = hashlib.md5(job_url.encode("utf-8")).hexdigest()[:12]
    if job_id != f"scraped-{CONFIG['trade']}-{digest}":
        return "id-url-mismatch"
    source = job.get("source")
    if source is not None and (
        not isinstance(source, str) or source.casefold() not in CONFIG["allowedSources"]
    ):
        return "invalid-source"
    return None


def profile_for(profile: str) -> dict[str, list[str]]:
    return CONFIG["profiles"][profile]
